flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/BinaryArray.java

```diff
@@ -361,6 +378,32 @@ public void setDecimal(int pos, Decimal value, int precision) {
 		}
 	}
 
+	@Override
+	public void setTimestamp(int pos, SqlTimestamp value, int precision) {
+		assertIndexIsValid(pos);
+
+		if (SqlTimestamp.isCompact(precision)) {
+			setLong(pos, value.getMillisecond());
+		} else {
+			int fieldOffset = getElementOffset(pos, 8);
+			int cursor = (int) (SegmentsUtil.getLong(segments, fieldOffset) >>> 32);
+			assert cursor > 0 : "invalid cursor " + cursor;
+
+			// zero-out the bytes
+			SegmentsUtil.setLong(segments, offset + cursor, 0L);
```

Comment:
move this setLong to `if (value == null)`. This is not decimal, timestamp is fix-length.
Others format too.